packages/frontend/core/src/components/attachment-viewer/worker/worker.ts

```diff
@@ -0,0 +1,223 @@
+import { DebugLogger } from '@affine/debug';
+import type { Document } from '@toeverything/pdf-viewer';
+import {
+  createPDFium,
+  PageRenderingflags,
+  Runtime,
+  Viewer,
+} from '@toeverything/pdf-viewer';
+
+import type { MessageData, MessageDataType } from './types';
+import { MessageOp, RenderKind } from './types';
+
+const logger = new DebugLogger('affine:worker:pdf');
+
+let inited = false;
+let viewer: Viewer | null = null;
+let doc: Document | undefined = undefined;
+
+// Caches images with the range.
+const cached = new Map<string, ImageData>();
+const docInfo = { total: 0, width: 1, height: 1 };
+const flags = PageRenderingflags.REVERSE_BYTE_ORDER | PageRenderingflags.ANNOT;
+const ranges = {
+  [`${RenderKind.Page}:startIndex`]: 0,
+  [`${RenderKind.Page}:endIndex`]: 0,
+  [`${RenderKind.Thumbnail}:startIndex`]: 0,
+  [`${RenderKind.Thumbnail}:endIndex`]: 0,
+};
+
+function post<T extends MessageOp>(type: T, data?: MessageDataType[T]) {
+  const message = { type, [type]: data };
+  self.postMessage(message);
+}
+
+function renderToImageData(index: number, scale: number) {
+  if (!viewer || !doc) return;
+
+  const page = doc.page(index);
+
+  if (!page) return;
+
+  const width = Math.ceil(docInfo.width * scale);
+  const height = Math.ceil(docInfo.height * scale);
+
+  const bitmap = viewer.createBitmap(width, height, 0);
+  bitmap.fill(0, 0, width, height);
+  page.render(bitmap, 0, 0, width, height, 0, flags);
+
+  const data = new Uint8ClampedArray(bitmap.toBytes());
+
+  bitmap.close();
+  page.close();
+
+  return new ImageData(data, width, height);
+}
+
+function createJob(index: number, kind: RenderKind, scale: number) {
+  return () => runJob(index, kind, scale);
+}
+
+async function runJob(index: number, kind: RenderKind, scale: number) {
+  const key = `${kind}:${index}`;
+
+  let imageData = cached.size > 0 ? cached.get(key) : undefined;
+
+  if (!imageData) {
+    try {
+      imageData = renderToImageData(index, scale);
+    } catch (err) {
+      console.error(err);
+    }
+
+    if (!imageData) return;
+
+    cached.set(key, imageData);
```

Comment:
why we need cache here, it seems that this cache will be cleared when all jobs done